src/toolchain/mod.rs

```diff
@@ -70,8 +92,36 @@ pub async fn download_file(
             .map_err(|_| Error::CreateDirectory(output_directory.to_string()))?;
     }
     info!("Downloading '{}'", &file_name);
-    let resp = reqwest::get(&url).await?;
-    let bytes = resp.bytes().await?;
+
+    let resp = {
+        let mut builder = reqwest::Client::builder();
+        if let Some(proxy) = https_proxy() {
+            builder = builder.proxy(reqwest::Proxy::https(&proxy).unwrap());
+        }
+        let client = builder.build()?;
+        client.get(&url).send().await?
+    };
+    let bytes = {
+        let len = resp.content_length();
+        let mut size_downloaded = 0;
+        let mut stream = resp.bytes_stream();
+        let mut bytes = bytes::BytesMut::new();
+        while let Some(chunk_result) = stream.next().await {
+            let chunk = chunk_result?;
+            size_downloaded += chunk.len();
+            if let Some(len) = len {
+                print!(
+                    "\rDownloading {file_name} {}/{} bytes",
+                    size_downloaded, len
+                );
+            } else {
+                print!("\rDownloaded {file_name} {} bytes", size_downloaded);
```

Comment:
I would suggest using `info` instead of `print`s.
```suggestion
                info!(
                    "\rDownloading {file_name} {}/{} bytes",
                    size_downloaded, len
                );
            } else {
                info!("\rDownloaded {file_name} {} bytes", size_downloaded);
```

Also, I noticed that the logs are wrongly shown (since multiple files are being downloaded at the same time ) and then wrongly formatted:
```
Downloading rust.tar.xz[info]: Creating symlink between '/home/esp/.rustup/toolchains/esp/xtensa-esp32-elf-clang/esp-18.1.2_20240912/esp-clang/lib' and '/home/esp/.espup/esp-clang'
Downloading [info]: Installing 'rust' component for Xtensa Rust toolchain
[info]: Downloading 'rust-src.tar.xz'
Downloading rust-src.tar.xz 3029727/[info]: Installing 'rust-src' component for Xtensa Rust toolchain
[info]: Installation successfully completed!
Downloading rust-src.tar.xz 3196644/3196644 bytes
```